File: ios/EatPal/EatPal/Services/GroceryTextParser.swift
import Foundation

/// Lightweight Swift port of `src/lib/parse-grocery-text.ts`.
/// Optimised for spoken input (voice-to-grocery, US-140) and pasted free text
/// (screenshot OCR, US-141). Outputs `ParsedGroceryItem` values ready to be
/// materialised into `GroceryItem` rows.
///
/// Intentionally simpler than the TS original — regex-heavy quantity parsing
/// isn't needed for the speech path because SFSpeechRecognizer already
/// normalises numerals ("two pounds of chicken" → "2 pounds of chicken").
struct ParsedGroceryItem: Equatable, Identifiable {
    let id = UUID()
    var name: String
    var quantity: Double
    var unit: String
    var category: String  // FoodCategory.rawValue

    /// Approximate confidence in the parse, 0.0–1.0. Low confidence items can
    /// be surfaced with a review prompt before insert.
    var confidence: Double
}

enum GroceryTextParser {
    // MARK: - Dictionaries

    private static let categoryKeywords: [(category: String, keywords: [String])] = [
        ("protein", [
            "chicken", "beef", "pork", "turkey", "salmon", "tuna", "shrimp", "fish",
            "steak", "ground beef", "bacon", "sausage", "ham", "lamb", "tofu",
            "eggs", "egg", "meatball", "hot dog", "deli", "jerky", "bison",
            "tilapia", "cod", "crab", "lobster", "scallop", "clam", "mussel"
        ]),
        ("dairy", [
            "milk", "cheese", "yogurt", "butter", "cream", "sour cream", "cottage cheese",
            "mozzarella", "cheddar", "parmesan", "ricotta", "cream cheese",
            "half and half", "ice cream", "gelato", "kefir", "ghee"
        ]),
        ("fruit", [
            "apple", "banana", "orange", "grape", "strawberr", "blueberr", "raspberr",
            "blackberr", "mango", "pineapple", "watermelon", "cantaloupe", "honeydew",
            "peach", "pear", "plum", "cherry", "kiwi", "lemon", "lime", "avocado",
            "coconut", "pomegranate", "fig", "papaya", "cranberr", "melon"
        ]),
        ("vegetable", [
            "broccoli", "carrot", "spinach", "kale", "lettuce", "tomato", "cucumber",
            "pepper", "onion", "garlic", "potato", "corn", "pea",
            "green bean", "celery", "mushroom", "zucchini", "squash", "cauliflower",
            "asparagus", "beet", "cabbage", "eggplant", "radish", "leek", "arugula",
            "cilantro", "parsley", "basil", "ginger", "scallion", "shallot", "salad"
        ]),
        ("carb", [
            "bread", "rice", "pasta", "noodle", "tortilla", "bagel", "roll", "bun",
            "cereal", "oat", "oatmeal", "granola", "flour", "cracker", "chip",
            "pita", "wrap", "couscous", "quinoa", "barley", "waffle",
            "english muffin", "croissant", "biscuit", "breadcrumb", "pretzel", "popcorn"
        ]),
        ("snack", [
            "cookie", "candy", "chocolate", "gummy", "granola bar", "pudding",
            "trail mix", "nut", "almond", "peanut butter", "jelly", "jam", "honey",
            "syrup", "ketchup", "mustard", "mayo", "ranch", "salsa", "hummus",
            "dressing", "sauce", "juice", "soda", "coffee", "tea", "kombucha"
        ])
    ]

    private static let unitMap: [String: String] = [
        "lb": "lbs", "lbs": "lbs", "pound": "lbs", "pounds": "lbs",
        "oz": "oz", "ounce": "oz", "ounces": "oz",
        "g": "g", "gram": "g", "grams": "g",
        "kg": "kg", "kilo": "kg", "kilos": "kg", "kilogram": "kg", "kilograms": "kg",
        "gal": "gal", "gallon": "gal", "gallons": "gal",
        "qt": "qt", "quart": "qt", "quarts": "qt",
        "pt": "pt", "pint": "pt", "pints": "pt",
        "cup": "cups", "cups": "cups",
        "tbsp": "tbsp", "tablespoon": "tbsp", "tablespoons": "tbsp",
        "tsp": "tsp", "teaspoon": "tsp", "teaspoons": "tsp",
        "l": "L", "liter": "L", "liters": "L",
        "ml": "ml", "milliliter": "ml", "milliliters": "ml",
        "bunch": "bunch", "bunches": "bunch",
        "bag": "bags", "bags": "bags",
        "box": "boxes", "boxes": "boxes",
        "can": "cans", "cans": "cans",
        "jar": "jars", "jars": "jars",
        "bottle": "bottles", "bottles": "bottles",
        "pack": "packs", "packs": "packs", "package": "packs", "packages": "packs",
        "dozen": "dozen",
        "loaf": "loaves", "loaves": "loaves",
        "stick": "sticks", "sticks": "sticks"
    ]

    /// Word-form numbers commonly produced by SFSpeechRecognizer.
    private static let numberWords: [String: Double] = [
        "a": 1, "an": 1, "one": 1, "two": 2, "three": 3, "four": 4, "five": 5,
        "six": 6, "seven": 7, "eight": 8, "nine": 9, "ten": 10,
        "eleven": 11, "twelve": 12, "half": 0.5, "quarter": 0.25
    ]

    private static let sentinelTokens: Set<String> = ["of", "and", "also", "plus", "some"]

    // MARK: - Public API

    /// Parse freeform text into grocery items. Accepts comma-separated, newline-separated,
    /// or spoken phrases joined by " and " / " also ".
    static func parse(_ text: String) -> [ParsedGroceryItem] {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return [] }

        // Split on common separators. Prefer newlines > commas > " and " / " also ".
        let separators: CharacterSet = CharacterSet(charactersIn: "\n,;")
        var phrases = trimmed.components(separatedBy: separators)

        // If we didn't split at all (single utterance), try word-level conjunctions.
        if phrases.count == 1 {
            let connectorSplit = trimmed
                .replacingOccurrences(of: " and also ", with: ",")
                .replacingOccurrences(of: " and then ", with: ",")
                .replacingOccurrences(of: " and ", with: ",")
                .replacingOccurrences(of: " also ", with: ",")
                .replacingOccurrences(of: " plus ", with: ",")
                .components(separatedBy: ",")
            if connectorSplit.count > 1 {
                phrases = connectorSplit
            }
        }

        var items: [ParsedGroceryItem] = []
        var seenNames: Set<String> = []

        for rawPhrase in phrases {
            guard let parsed = parsePhrase(rawPhrase) else { continue }
            let key = parsed.name.lowercased()
            if seenNames.contains(key) { continue }
            seenNames.insert(key)
            items.append(parsed)
        }

        return items
    }

    // MARK: - Phrase parser

    private static func parsePhrase(_ phrase: String) -> ParsedGroceryItem? {
        // Clean leading bullets / numbers / checkboxes so we handle pasted lists too.
        var cleaned = phrase.trimmingCharacters(in: .whitespacesAndNewlines)
        cleaned = stripListMarkers(cleaned)
        guard cleaned.count >= 2 else { return nil }

        var tokens = cleaned
            .lowercased()
            .components(separatedBy: .whitespacesAndNewlines)
            .filter { !$0.isEmpty }

        guard !tokens.isEmpty else { return nil }

        var quantity: Double = 1
        var quantityMatched = false
        var unit = ""
        var confidence: Double = 0.6

        // 1) First token: numeric quantity or number word.
        if let firstNumeric = parseNumber(tokens[0]) {
            quantity = firstNumeric
            quantityMatched = true
            tokens.removeFirst()
            confidence = 0.8
        }

        // 2) Drop common filler like "of", "some".
        while let first = tokens.first, sentinelTokens.contains(first) {
            tokens.removeFirst()
        }

        // 3) Next token might be a unit.
        if let first = tokens.first, let normalizedUnit = unitMap[first] {
            unit = normalizedUnit
            tokens.removeFirst()
            if !quantityMatched { quantity = 1 }
            confidence = max(confidence, 0.85)
        }

        // 4) Drop filler again ("two pounds OF chicken").
        while let first = tokens.first, sentinelTokens.contains(first) {
            tokens.removeFirst()
        }

        guard !tokens.isEmpty else { return nil }

        let name = tokens.joined(separator: " ")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        guard name.count >= 2 else { return nil }

        let category = inferCategory(name)

        // If we never matched quantity or unit and the name is short, drop confidence.
        if !quantityMatched && unit.isEmpty && name.split(separator: " ").count == 1 {
            confidence = 0.7
        }

        return ParsedGroceryItem(
            name: titleCase(name),
            quantity: quantity,
            unit: unit,
            category: category,
            confidence: confidence
        )
    }

    // MARK: - Helpers

    private static func stripListMarkers(_ input: String) -> String {
        var s = input

        // Leading bullets / dashes
        while let first = s.first, "-•*▪▸►◆☐☑✓✔".contains(first) {
            s.removeFirst()
        }

        // "1.", "1)"
        if let match = s.range(of: #"^\s*\d+[\.\)]\s*"#, options: .regularExpression) {
            s.removeSubrange(match)
        }

        // "[ ]" / "[x]"
        if let match = s.range(of: #"^\s*\[[ xX]?\]\s*"#, options: .regularExpression) {
            s.removeSubrange(match)
        }

        return s.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private static func parseNumber(_ token: String) -> Double? {
        // Plain number / decimal
        if let value = Double(token), value > 0 { return value }

        // Word form
        if let value = numberWords[token] { return value }

        // Fraction "1/2", "3/4"
        let parts = token.split(separator: "/")
        if parts.count == 2,
           let num = Double(parts[0]),
           let den = Double(parts[1]),
           den > 0 {
            return num / den
        }

        return nil
    }

    private static func inferCategory(_ name: String) -> String {
        let lower = name.lowercased()
        for (category, keywords) in categoryKeywords {
            if keywords.contains(where: { lower.contains($0) }) {
                return category
            }
        }
        return "snack"
    }

    private static func titleCase(_ s: String) -> String {
        s.split(separator: " ")
            .map { $0.prefix(1).uppercased() + $0.dropFirst() }
            .joined(separator: " ")
    }
}

File: android-native/app/src/main/java/com/eatpal/app/domain/GroceryTextParser.kt
package com.eatpal.app.domain

import java.util.UUID

/**
 * Kotlin port of iOS `GroceryTextParser`. Same dictionaries, same split
 * strategy, same confidence heuristics — designed for voice-to-grocery and
 * pasted OCR text. Shared output shape so iOS / Android review UIs stay
 * identical.
 */
object GroceryTextParser {

    data class ParsedItem(
        val id: String = UUID.randomUUID().toString(),
        val name: String,
        val quantity: Double,
        val unit: String,
        val category: String,
        val confidence: Double,
    )

    private val categoryKeywords: List<Pair<String, List<String>>> = listOf(
        "protein" to listOf(
            "chicken", "beef", "pork", "turkey", "salmon", "tuna", "shrimp", "fish",
            "steak", "ground beef", "bacon", "sausage", "ham", "lamb", "tofu",
            "eggs", "egg", "meatball", "hot dog", "deli", "jerky", "bison",
            "tilapia", "cod", "crab", "lobster", "scallop", "clam", "mussel",
        ),
        "dairy" to listOf(
            "milk", "cheese", "yogurt", "butter", "cream", "sour cream", "cottage cheese",
            "mozzarella", "cheddar", "parmesan", "ricotta", "cream cheese",
            "half and half", "ice cream", "gelato", "kefir", "ghee",
        ),
        "fruit" to listOf(
            "apple", "banana", "orange", "grape", "strawberr", "blueberr", "raspberr",
            "blackberr", "mango", "pineapple", "watermelon", "cantaloupe", "honeydew",
            "peach", "pear", "plum", "cherry", "kiwi", "lemon", "lime", "avocado",
            "coconut", "pomegranate", "fig", "papaya", "cranberr", "melon",
        ),
        "vegetable" to listOf(
            "broccoli", "carrot", "spinach", "kale", "lettuce", "tomato", "cucumber",
            "pepper", "onion", "garlic", "potato", "corn", "pea",
            "green bean", "celery", "mushroom", "zucchini", "squash", "cauliflower",
            "asparagus", "beet", "cabbage", "eggplant", "radish", "leek", "arugula",
            "cilantro", "parsley", "basil", "ginger", "scallion", "shallot", "salad",
        ),
        "carb" to listOf(
            "bread", "rice", "pasta", "noodle", "tortilla", "bagel", "roll", "bun",
            "cereal", "oat", "oatmeal", "granola", "flour", "cracker", "chip",
            "pita", "wrap", "couscous", "quinoa", "barley", "waffle",
            "english muffin", "croissant", "biscuit", "breadcrumb", "pretzel", "popcorn",
        ),
        "snack" to listOf(
            "cookie", "candy", "chocolate", "gummy", "granola bar", "pudding",
            "trail mix", "nut", "almond", "peanut butter", "jelly", "jam", "honey",
            "syrup", "ketchup", "mustard", "mayo", "ranch", "salsa", "hummus",
            "dressing", "sauce", "juice", "soda", "coffee", "tea", "kombucha",
        ),
    )

    private val unitMap: Map<String, String> = mapOf(
        "lb" to "lbs", "lbs" to "lbs", "pound" to "lbs", "pounds" to "lbs",
        "oz" to "oz", "ounce" to "oz", "ounces" to "oz",
        "g" to "g", "gram" to "g", "grams" to "g",
        "kg" to "kg", "kilo" to "kg", "kilos" to "kg", "kilogram" to "kg", "kilograms" to "kg",
        "gal" to "gal", "gallon" to "gal", "gallons" to "gal",
        "qt" to "qt", "quart" to "qt", "quarts" to "qt",
        "pt" to "pt", "pint" to "pt", "pints" to "pt",
        "cup" to "cups", "cups" to "cups",
        "tbsp" to "tbsp", "tablespoon" to "tbsp", "tablespoons" to "tbsp",
        "tsp" to "tsp", "teaspoon" to "tsp", "teaspoons" to "tsp",
        "l" to "L", "liter" to "L", "liters" to "L",
        "ml" to "ml", "milliliter" to "ml", "milliliters" to "ml",
        "bunch" to "bunch", "bunches" to "bunch",
        "bag" to "bags", "bags" to "bags",
        "box" to "boxes", "boxes" to "boxes",
        "can" to "cans", "cans" to "cans",
        "jar" to "jars", "jars" to "jars",
        "bottle" to "bottles", "bottles" to "bottles",
        "pack" to "packs", "packs" to "packs", "package" to "packs", "packages" to "packs",
        "dozen" to "dozen",
        "loaf" to "loaves", "loaves" to "loaves",
        "stick" to "sticks", "sticks" to "sticks",
    )

    private val numberWords: Map<String, Double> = mapOf(
        "a" to 1.0, "an" to 1.0, "one" to 1.0, "two" to 2.0, "three" to 3.0, "four" to 4.0,
        "five" to 5.0, "six" to 6.0, "seven" to 7.0, "eight" to 8.0, "nine" to 9.0, "ten" to 10.0,
        "eleven" to 11.0, "twelve" to 12.0, "half" to 0.5, "quarter" to 0.25,
    )

    private val sentinelTokens = setOf("of", "and", "also", "plus", "some")

    fun parse(text: String): List<ParsedItem> {
        val trimmed = text.trim()
        if (trimmed.isEmpty()) return emptyList()

        var phrases = trimmed.split('\n', ',', ';').map { it.trim() }.filter { it.isNotBlank() }
        if (phrases.size == 1) {
            val connectorSplit = trimmed
                .replace(" and also ", ",")
                .replace(" and then ", ",")
                .replace(" and ", ",")
                .replace(" also ", ",")
                .replace(" plus ", ",")
                .split(',').map { it.trim() }.filter { it.isNotBlank() }
            if (connectorSplit.size > 1) phrases = connectorSplit
        }

        val seen = mutableSetOf<String>()
        val items = mutableListOf<ParsedItem>()
        for (raw in phrases) {
            val parsed = parsePhrase(raw) ?: continue
            val key = parsed.name.lowercase()
            if (key in seen) continue
            seen += key
            items += parsed
        }
        return items
    }

    private fun parsePhrase(phrase: String): ParsedItem? {
        var cleaned = stripListMarkers(phrase.trim())
        if (cleaned.length < 2) return null

        val tokens = cleaned.lowercase().split(Regex("\\s+")).filter { it.isNotBlank() }.toMutableList()
        if (tokens.isEmpty()) return null

        var quantity = 1.0
        var quantityMatched = false
        var unit = ""
        var confidence = 0.6

        parseNumber(tokens.first())?.let {
            quantity = it
            quantityMatched = true
            tokens.removeAt(0)
            confidence = 0.8
        }

        while (tokens.isNotEmpty() && tokens.first() in sentinelTokens) tokens.removeAt(0)

        tokens.firstOrNull()?.let { first ->
            unitMap[first]?.let {
                unit = it
                tokens.removeAt(0)
                if (!quantityMatched) quantity = 1.0
                confidence = maxOf(confidence, 0.85)
            }
        }

        while (tokens.isNotEmpty() && tokens.first() in sentinelTokens) tokens.removeAt(0)
        if (tokens.isEmpty()) return null

        val name = tokens.joinToString(" ").trim()
        if (name.length < 2) return null

        if (!quantityMatched && unit.isEmpty() && name.split(" ").size == 1) confidence = 0.7

        return ParsedItem(
            name = titleCase(name),
            quantity = quantity,
            unit = unit,
            category = inferCategory(name),
            confidence = confidence,
        )
    }

    private fun stripListMarkers(input: String): String {
        var s = input
        while (s.isNotEmpty() && s.first() in "-•*▪▸►◆☐☑✓✔") s = s.substring(1)
        s = s.replace(Regex("^\\s*\\d+[.)]\\s*"), "")
        s = s.replace(Regex("^\\s*\\[[ xX]?]\\s*"), "")
        return s.trim()
    }

    private fun parseNumber(token: String): Double? {
        token.toDoubleOrNull()?.let { if (it > 0) return it }
        numberWords[token]?.let { return it }
        val parts = token.split("/")
        if (parts.size == 2) {
            val num = parts[0].toDoubleOrNull()
            val den = parts[1].toDoubleOrNull()
            if (num != null && den != null && den > 0) return num / den
        }
        return null
    }

    private fun inferCategory(name: String): String {
        val lower = name.lowercase()
        for ((category, keywords) in categoryKeywords) {
            if (keywords.any { lower.contains(it) }) return category
        }
        return "snack"
    }

    private fun titleCase(s: String): String =
        s.split(" ").joinToString(" ") { w -> w.replaceFirstChar { it.titlecase() } }
}
